web socket enhancement

Hammad887 opened the pull request:
1. Implemented Web Socket for real time notifications.
2. Unit tests written for the web socket package
3. Made some changes in swagger.yaml and db layer (Found some issues with CRUD operations while testing)
4. Implemented Validation Checks to make sure only authenticated users can enter and send message to a particular chatroom.

Hammad887 committed:
websocket enhancement

shahzadhaider1 commented on cmd/chat_app_server/main.go:
```diff
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"github.com/Hammad887/chat-app/websocket"
```
sort imports

Hammad887 replied:
sorted

shahzadhaider1 replied:
here is the correct sort order of imports:
1. Standard Library Packages: Import standard library packages first.
2. Third-Party Packages: Import external (third-party) packages second. These are packages not part of the standard library.
3. Local Packages: Import your own local packages third. These are packages that are part of your project.

Hammad887 replied:
I read somewhere that goimports automatically does that for you. I just realized that goimports only separates standard library imports from third-party imports, but does not differ between third-party imports and own local imports. Sorted now

shahzadhaider1 commented on db/mysql/ops.go:
```diff
@@ -59,11 +59,25 @@ func (c *client) ListChatRoom(ctx context.Context) ([]*domain.ChatRoom, error) {
 	}
 
 	return &chatRoom, nil
+
```
no need for this blank line

Hammad887 replied:
deleted

shahzadhaider1 commented on swagger.yaml:
```diff
@@ -47,6 +47,37 @@ paths:
+            $ref: '#/definitions/ChatroomInfo'
+      responses:
+        200:
+          description: Successful created a chatroom
```
Successfully created a chatroom

Hammad887 replied:
changed

shahzadhaider1 commented on swagger.yaml:
```diff
@@ -47,6 +47,37 @@ paths:
+      description: 'Creates a new chatroom with the provide name'
+      tags:
+        - service
+      operationId: ChatRoom
```
For operationId, you should ideally use a verb to describe the action being performed. For example, you could use CreateChatRoom here

Hammad887 replied:
done

shahzadhaider1 commented on swagger.yaml:
```diff
@@ -47,6 +47,37 @@ paths:
+          description: Successful created a chatroom
+          schema:
+            $ref: '#/definitions/ChatroomSuccess'
+        400:
+          description: Bad Request
```
the `parameter` in this API is `required`, so the request won’t be allowed if the body is missing/incorrect. 
are there any other cases where this 400 response could be utilised?

Hammad887 replied:
Can't think of anything on top of my head right now. Deleted 400 for now.

shahzadhaider1 commented on swagger.yaml:
```diff
@@ -47,6 +47,37 @@ paths:
+        404:
+          schema:
+            type: string
+          description: User not found
```
User not found?

Hammad887 replied:
changed

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+	"strconv"
+
+	"github.com/Hammad887/chat-app/db"
+	"github.com/Hammad887/chat-app/db/mysql"
```
sort imports

Hammad887 replied:
sorted

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+type ChatRoom struct {
+	chatRoomId string
+	clients    []*Clients
+}
```
we have a struct with the same name in the models package. why that wasn’t used here?

Hammad887 replied:
I have integrated some changes in the models package and using it here now without making a new struct

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+		TestMode: false, // Set the appropriate value for TestMode here
+	}
+
+	DbClient, _ = mysql.NewClient(options)
```
we have already created a database client in runtime file, right? 
why that wasn’t used here?
why multiple database clients in a single application?

Hammad887 replied:
My bad. Was blindly following web implementations.. Made a minor change in main.go file. Now only one db client in the application.

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+	},
+}
+
+func RunWebsocketServer() {
```
add comments for exported functions

Hammad887 replied:
Added

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+	http.ListenAndServe(":8085", router)
+}
+
+func ClientSocket(w http.ResponseWriter, r *http.Request) {
```
add comments for exported functions

Hammad887 replied:
Added

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+		messageType, requestByte, err := Conn.ReadMessage()
+
+		if err != nil {
+			log.Println("ERROR read", err)
```
use Error() here

Hammad887 replied:
Done

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+			if _, ok := requestJson["sender_id"]; ok {
+				if !ok {
+					Response := make(map[string]interface{})
+					Response["request_type"] = "connect_to_chat_room"
```
I see "connect_to_chat_room" and other magic values used in multiple places. 
Consider using constants for these.

Hammad887 replied:
Done

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+
+}
+
+func WriteOnSocket(responseBytes []byte, messageType int, Conn *websocket.Conn) {
```
add comments for exported functions

Hammad887 replied:
added

shahzadhaider1 commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,295 @@
+	}
+}
+
+// InterfaceToString (interface{}) get value of an interface return as string {
+func InterfaceToString(ThisInterface interface{}) string {
+	if ThisInterface == nil {
+		fmt.Println("InterfaceToString :: NIL value passed for conversion")
+		return ""
+	}
+	switch ThisInterface.(type) {
+	case int:
+		return strconv.Itoa(ThisInterface.(int))
+	case int64:
+		return strconv.FormatInt(ThisInterface.(int64), 10)
+	case float64:
+		TmpStr := strconv.FormatFloat(ThisInterface.(float64), 'f', 10, 64)
+		if TmpStr[(len(TmpStr)-10):] == "0000000000" {
+			return TmpStr[:(len(TmpStr) - 11)]
+		}
+		return TmpStr
+	default:
+		return ThisInterface.(string)
+	}
+}
```
this function converts interface to multiple types, why only use string in the function name?

Hammad887 replied:
This function was unnecessary. Have deleted it.

shahzadhaider1 commented on websocket/websocket-server_test.go:
```diff
@@ -0,0 +1,47 @@
+	"testing"
+
+	"github.com/gorilla/websocket"
+
```
no need for this space

Hammad887 replied:
deleted

shahzadhaider1 commented on websocket/websocket-server_test.go:
```diff
@@ -0,0 +1,47 @@
+	}
+	defer conn.Close()
+
+}
```
why tests for only two functions?

Hammad887 replied:
Tests added for all the functions

shahzadhaider1 commented on models/user.go:
```diff
@@ -1,9 +1,12 @@
+	Name       string          `json:"name"`
+	Email      string          `json:"email"`
+	Password   string          `json:"-"`
+	Connection *websocket.Conn `json:"connection,omitempty"`
```
why not create a new struct in your model layer that represents a WebSocket connection, which includes the "Connection *websocket.Conn" field. This struct can be something like `WebSocketConnection`.

shahzadhaider1 replied:
I want you to take a moment and think about this and come up with an optimal approach (according to you)

Hammad887 replied:
The best approach I could think of after some research as well was to make separate structs for both user and chatroom connection field. Pushed the latest for review.

Hammad887 commented:
I have uploaded a small video to demo the functionality. Here's the link: https://drive.google.com/file/d/1l31IjJrvoWk9nmKk_Wq9_xtkHosbXgx2/view?usp=sharing

shahzadhaider1 commented:
Hammad, that's impressive. 
1. But I didn’t see any authentication being performed at the beginning. You didn’t provide any credentials, etc. did you do it before making the video? 
2. Could you create a third user and connect that user to a different chatroom and then send the message? Now, this third user shouldn’t receive the message since this user is not a part of the chatroom where the message was sent.

shahzaibaziz commented on models/chatroom.go:
```diff
@@ -2,7 +2,11 @@ package models
+	ID                 string   `json:"id"`                           // to hold the ID of User
+	Name               string   `json:"name"`                         // to hold the Name of User
+	Users              []string `json:"users"`                        // to hold array of ID of Users
+	ClientWsConnection []*User  `json:"ClientWsConnection,omitempty"` // to hold array of User's complete struct information with websocket connection address
```
```suggestion
	ClientWsConnection []*User  `json:"client_ws_connection,omitempty"` // to hold array of User's complete struct information with websocket connection address
```

Hammad887 replied:
Done

shahzaibaziz commented on cmd/chat_app_server/main.go:
```diff
@@ -29,6 +30,8 @@ func main() {
 
 	handler.NewCustomHandler(api, rt)
 
+	go websocket.RunWebsocketServer(rt)
+
```
Instead of creating a new server, have you considered adding this as middleware in this file?https://github.com/Hammad887/chat-app/blob/master/gen/restapi/configure_chatroom.go#L98

Hammad887 replied:
Looking into this

Hammad887 replied:
Good to go

shahzaibaziz replied:
We need to create a separate main function to run the WebSocket server within the cmd folder, specifically in chat-app-websocket/main.go. Additionally, we should establish a different entry point in the Docker service in your docker-compose file. This will involve defining two separate services, one for the chat server and the other for the WebSocket server.

Hammad887 replied:
Created a separate main function to run the WebSocket server. Will discuss about Dockerfile and docker-compose file during a call.

Hammad887 commented:
@shahzadhaider1 Yes, I did the authentication before making the video. Also, here's another video for the 3rd user you wanted me to create and test with the previously made chatroom. Link: https://drive.google.com/file/d/11YbK8n_fTDpPK8bKAxlg9wGoZyLLb9RB/view?usp=sharing

shahzadhaider1 commented:
Alright, good work.

shahzaibaziz commented on models/websocket-user.go:
```diff
@@ -0,0 +1,9 @@
+// WebsocketUser holds information for a WebsocketUser user
+type WebsocketUser struct {
+	ID         string          `json:"id"`
+	Connection *websocket.Conn `json:"connection,omitempty"`
```
can you please explain why we are storing the connection string to the database?

Hammad887 replied:
I read online that it's a good practice for microservice architecture. It is basically used for secure communication and service-to-service authorization/authentication. For example, it will ensure that a user can only connect to a WebSocket if they have a valid token.

shahzaibaziz commented on swagger.yaml:
```diff
@@ -47,6 +47,32 @@ paths:
             type: string
           description: Server error
 
+  /createChatRoom:
```
It should be 'create-chat-room,' which follows the basic convention used in APIs

Hammad887 replied:
changed

shahzaibaziz commented on websocket/websocket-enums.go:
rename this file to constant

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-enums.go:
```diff
@@ -0,0 +1,11 @@
+
+//Defines Websocket enums
+
+const RequestTypeChatRoom = "connect_to_chat_room"
+const RequestTypeSendMessage = "send_message"
+const RequestParamRoomId = "room_id"
+const RequestParamSenderId = "sender_id"
+const RequestParamMessageFrom = "message_from"
+const RequestParamMessage = "message"
+const RequestParamRequestType = "requestType"
```
```suggestion
const (
 RequestTypeChatRoom = "connect_to_chat_room"
 RequestTypeSendMessage = "send_message"
RequestParamRoomId = "room_id"
RequestParamSenderId = "sender_id"
RequestParamMessageFrom = "message_from"
RequestParamMessage = "message"
RequestParamRequestType = "requestType"
)
```
Arrange them in alphabetical order

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+)
+
+// ctx is a variable of type context.Context, which is used to manage request-scoped values, deadlines, and cancellations.
+var ctx context.Context
```
why we are using this context here, it should be passed from the main file , like we are using the swagger server

Hammad887 replied:
Using context.Background() in the main file.

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+	// EnableCompression allows for message compression when set to true.
+	EnableCompression: true,
+
+	// CheckOrigin is a function that checks the origin of the WebSocket request to prevent cross-origin attacks.
+	// In this case, it's set to always return true, meaning any origin is accepted.
+	CheckOrigin: func(r *http.Request) bool {
+		return true
+	},
+}
```
Please remove the whole  logic from your server. In an ideal situation, microservices are not publicly exposed directly to the internet. Instead, we typically use a server-client proxy (e.g., Nginx) that serves as the public-facing entry point to your server

Hammad887 replied:
Removed

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+	router.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
+		ClientSocket(w, r, rt)
+	})
+	fmt.Println("Server is running on port 8085")
```
it should be and thus logic move to main file

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+	fmt.Println("Server is running on port 8085")
+
+	// Start your server
+	go func() {
+		err := http.ListenAndServe(":8085", router)
+		if err != nil {
+			log.Println(err)
+		}
+	}()
```
move to main file

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+
+	Conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Print("upgrade:", err)
```
handle error correctly, it should pass to the  handler and return error message

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+		}
+	}(Conn)
+
+	rt.Service()
```
remove this

Hammad887 replied:
Removed

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+	}(Conn)
+
+	rt.Service()
+	service := rt.Service()
```
this logic should be in runtime

shahzaibaziz replied:
you can use a flag list in NewRuntime function to create a different runtime for websocket server and chat server

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+		requestJson := make(map[string]interface{})
+		err = json.Unmarshal(requestByte, &requestJson)
+		if err != nil {
+			log.Println(err)
+			return
```
error handling issue

Hammad887 replied:
Resolved

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+
+			if _, ok := requestJson[RequestParamRoomId]; ok {
+				if !ok {
+					Response := make(map[string]interface{})
```
use this on top of switch

Hammad887 replied:
Moved

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+					Response := make(map[string]interface{})
+					Response[RequestParamRequestType] = RequestTypeChatRoom
+					Response[RequestParamMessage] = "required " + RequestParamRoomId
+					responseBytes, _ := json.Marshal(Response)
```
unhandled error

Hammad887 replied:
Resolved

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+					Response := make(map[string]interface{})
+					Response[RequestParamRequestType] = RequestTypeChatRoom
+					Response[RequestParamMessage] = "required " + RequestParamSenderId
+					responseBytes, _ := json.Marshal(Response)
```
unhandled error

Hammad887 replied:
Resolved

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+				Response := make(map[string]interface{})
+				Response[RequestParamRequestType] = RequestTypeChatRoom
+				Response[RequestParamMessage] = "required string type " + RequestParamRoomId
+				responseBytes, _ := json.Marshal(Response)
```
same as above

Hammad887 replied:
Resolved

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+			roomId, ok := requestJson[RequestParamRoomId].(string)
+
+			if !ok {
+				Response := make(map[string]interface{})
```
same as above

Hammad887 replied:
Resolved

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+
+			senderId, ok := requestJson[RequestParamSenderId].(string)
+			if !ok {
+				Response := make(map[string]interface{})
```
avoid creating a same variable in each case declare this on top of switch

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+				break
+			}
+
+			row, _ := service.GetChatroom(ctx, roomId)
```
handle error,

Hammad887 replied:
Done

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+	rt.Service()
+	service := rt.Service()
+
+	for {
```
why using infinite loop here?

Hammad887 replied:
I am using break statements. The loop will terminate when the connection breaks.

shahzaibaziz commented on websocket/websocket-server.go:
```diff
@@ -0,0 +1,335 @@
+			log.Println(err)
+			return
+		}
+		switch requestJson[RequestParamRequestType] {
```
One major thing is missing here: in the case of an error, the error message must be displayed in Postman, this message is related to all the cases in a switch

shahzaibaziz replied:
Reduce the redundancy in the code

Hammad887 replied:
All error messages are displaying in Postman.

Hammad887 replied:
Added a function at the end to reduce the redundancy

shahzaibaziz commented on websocket/websocket-server_test.go:
```diff
@@ -0,0 +1,164 @@
+		args args
+	}{
+		{
+			name: "CheckConnection",
+			args: args{
+				responseBytes: []byte(`{"requestType":"connect_socket"}`),
+				messageType:   websocket.TextMessage,
+				Conn:          conn, // Use the WebSocket connection to the server
+			},
+		},
+		{
+			name: "InvalidCheckConnection",
+			args: args{
+				responseBytes: []byte(`{"requestType": "connect_socket"}`),
+				messageType:   websocket.TextMessage,
+				Conn:          conn, // Use the WebSocket connection to the server
+			},
+		},
```
You don't need to change this, but I have a suggestion for your work. This test doesn't cover the entire server, and there's a significant lack of coverage for each case in your server

Hammad887 replied:
Will keep it under consideration in the future. Thank You!

shahzaibaziz commented:
Good work! Kudos to @Hammad887  for putting a lot of effort into it

Hammad887 commented:
@shahzaibaziz I have changed all the filenames to snake case. Thank you so much for imparting this knowledge. Also, added a websocket handler which addresses most of the changes you asked me to resolve in the websocket-server.go file in the previous commit.

shahzadhaider1 commented on docker-compose.yaml:
```diff
@@ -43,4 +65,4 @@ services:
     environment:
-      SWAGGER_JSON: /swagger.yaml
\ No newline at end of file
+      SWAGGER_JSON: /swagger.yaml
```
leave blank line at the end

Hammad887 replied:
Pushed the code again after adding the file, but it still isn't showing up here.

shahzaibaziz replied:
If you plan to provide an API specification page, you'll need to expose the port so that anyone who wants to view the API spec can do so by accessing the host and port to see the API documentation

shahzadhaider1 commented on docker-compose.yaml:
```diff
@@ -35,6 +37,26 @@ services:
+      # MYSQL_DB_HOSTS: mysql:3306
+      # MYSQL_DB_NAME: userdb
+      # MYSQL_DB_USERNAME: root
+      # MYSQL_DB_PASSWORD: password
```
remove these comments if not required

Hammad887 replied:
Removed.

shahzadhaider1 commented on handlers/websocket.go:
```diff
@@ -0,0 +1,301 @@
+	RequestParamRequestType = "requestType"
+	RequestParamRoomId      = "room_id"
+	RequestParamSenderId    = "sender_id"
+	RequestTypeChatRoom     = "connect_to_chat_room"
```
we can make it more descriptive

Hammad887 replied:
Using Prefix Constants and Clear Naming now.

shahzadhaider1 commented on models/chatroom.go:
```diff
@@ -2,7 +2,7 @@ package models
-	ID    string   `json:"id"`
-	Name  string   `json:"name"`
-	Users []string `json:"users"`
+	ID    string   `json:"id"`    // to hold the ID of User
+	Name  string   `json:"name"`  // to hold the Name of User
```
update the comments

Hammad887 replied:
updated

shahzadhaider1 commented on models/websocket_chatroom.go:
```diff
@@ -0,0 +1,10 @@
+
+// WebsocketChatRoom holds information for the users inside a particular chat room
+type WebsocketChatRoom struct {
+	ID                 string           `json:"id"`                             // to hold the ID of User
```
is this the ID of user?

Hammad887 replied:
My bad. Changed it to the ID of ChatRoom

shahzaibaziz commented on cmd/chat_app_socket_server/main.go:
```diff
@@ -0,0 +1,47 @@
+
+	rt, err := runtime.NewRuntime()
+	if err != nil {
+		log.Println(err)
```
```suggestion
		panic(err)
```

Hammad887 replied:
Changed

shahzaibaziz commented on cmd/chat_app_socket_server/main.go:
```diff
@@ -0,0 +1,47 @@
+	service := rt.Service()
+
+	router.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
+		err = handlers.WebSocketHandler(w, r, service, ctx)
```
paramter should be runtime object not service object

Hammad887 replied:
Changed

shahzaibaziz commented on cmd/chat_app_socket_server/main.go:
```diff
@@ -0,0 +1,47 @@
+	fmt.Println("Server is running on port 8085")
+	err = http.ListenAndServe(":8085", router)
+	if err != nil {
+		log.Println(err)
```
panic

Hammad887 replied:
Changed

shahzaibaziz commented on cmd/chat_app_socket_server/main.go:
```diff
@@ -0,0 +1,47 @@
+	})
+
+	fmt.Println("Server is running on port 8085")
+	err = http.ListenAndServe(":8085", router)
```
its better to use env variable for port number

Hammad887 replied:
Done

shahzaibaziz commented on docker-compose.yaml:
```diff
@@ -35,6 +37,26 @@ services:
+      PORT: 8085
+      # MYSQL_DB_HOSTS: mysql:3306
+      # MYSQL_DB_NAME: userdb
+      # MYSQL_DB_USERNAME: root
```
This Dockerfile change is also acceptable, but there is another way to build binary files during the 'make build' process and run binary files with the help of bash scripts. The consumption has the same implementation

Hammad887 replied:
Will look into this.

shahzaibaziz commented on handlers/sendMessage.go:
rename to send_message.go to consist with project

Hammad887 replied:
Done

shahzaibaziz reviewed (commented):
minor comments other than that LGTM

shahzaibaziz commented on .DS_Store:
remove this

Hammad887 replied:
Removed

shahzadhaider1 commented on cmd/chat_app_socket_server/main.go:
```diff
@@ -0,0 +1,56 @@
+// main.go
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+
+	"github.com/gorilla/mux"
+	"github.com/joho/godotenv"
+
+	runtime "github.com/Hammad887/chat-app"
+	"github.com/Hammad887/chat-app/handlers"
+	request "github.com/gorilla/handlers"
+)
```
please sort imports

Hammad887 replied:
Sorted

shahzadhaider1 reviewed (approved):
LGTM. Please resolve Shahzaib's comment and you’ll be good to merge.
GW!
